fix(agent/agentssh): use deterministic host key for SSH server

ThomasK33 opened the pull request:
Fixes: https://github.com/coder/coder/issues/16490

The Agent's SSH server now initially generates fixed host keys and, once it receives its manifest, generates and replaces that host key with the one derived from the workspace ID, ensuring consistency across agent restarts. This prevents SSH warnings and host key verification errors when connecting to workspaces through Coder Desktop.

While deterministic keys might seem insecure, the underlying Wireguard tunnel already provides encryption and anti-spoofing protection at the network layer, making this approach acceptable for our use case.

---
Change-Id: I8c7e3070324e5d558374fd6891eea9d48660e1e9
Signed-off-by: Thomas Kosiewski <tk@coder.com>

ThomasK33 committed:
fix(agent/agentssh): pin random seed for RSA key generation

Change-Id: I8c7e3070324e5d558374fd6891eea9d48660e1e9
Signed-off-by: Thomas Kosiewski <tk@coder.com>

ThomasK33 commented:
* **#16626** <a href="https://app.graphite.dev/github/pr/coder/coder/16626?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/coder/coder/16626?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -115,11 +115,15 @@ func NewServer(ctx context.Context, logger slog.Logger, prometheusRegistry *prom
+	// Create a deterministic random source
+	// nolint: gosec
+	deterministicRand := rand.New(rand.NewSource(42))
+	coderHostKey, err := rsa.GenerateKey(deterministicRand, 2048)
```
Could we base this on workspace and agent name? I'm wondering if there won't be conflicts between multiple workspaces having the same host key?

ThomasK33 replied:
Good point. I don't think there should be any issues, but I'll look into getting the names in as seed values.

ThomasK33 replied:
So, we need at least one host key to start the SSH server; thus, the hardcoded `42` seed. Later, once we receive the manifest, we use the workspace ID as a seed and derive a host key, replacing the host key derived from the `42` seed.

I am unsure if deriving the key from the workspace and agent names/UUIDs makes sense, as we need the host key fixed on a per-workspace basis. If we include the agent as a seed, connections using Coder Desktop's VPN would be presented with a different host key for the same workspace, thus still refusing to connect, assuming it's an impersonation.

mafredri replied:
The agent UUID will change every start, but the name should be stable. It is, however, possible to rename a workspace, but allowing that is a separate option that needs to be enabled. One option is `workspaceUUID.agentName`, which is as stable as it gets.

mafredri commented on agent/agentssh/agentssh.go:
```diff
@@ -1099,3 +1100,32 @@ func userHomeDir() (string, error) {
 }
+
+// UpdateHostSigner updates the host signer with a new key generated from the provided seed.
+// This function replaces the previously used host key with the newly generated one.
```
It could be useful to mention here that gliderlabs/ssh matches on key type and only allows one/type. At least for me I was wondering how this works and had to dive into the docs to find:

```
If an existing host key exists with the same algorithm, it is overwritten
```

ThomasK33 replied:
Yeah, the wording from `gliderlabs/ssh` is clearer.

mafredri reviewed (commented):
I think you found a great approach here, nice work! Just one minor nit wrt the seed source, otherwise LGTM.

I would still like to see a test that ensures the this behaves as expected.

spikecurtis commented on agent/agentssh/agentssh.go:
```diff
@@ -131,14 +131,15 @@ func NewServer(ctx context.Context, logger slog.Logger, prometheusRegistry *prom
+	// Create a fixed host key, as at least one host key has to
+	// exist for the SSH server to start. We'll later replace
+	// the host key with one that's based off the workspace uuid.
+	coderSigner, err := coderSigner(42)
```
Do we actually need a key at this time? I don't see the point of creating one if, by design, we replace it before anyone can possibly connect.

spikecurtis commented on agent/agent.go:
```diff
@@ -906,6 +906,14 @@ func (a *agent) handleManifest(manifestOK *checkpoint) func(ctx context.Context,
+		if err != nil {
+			return xerrors.Errorf("generate hash from workspace id: %w", err)
+		}
+		if err := a.sshServer.UpdateHostSigner(seed); err != nil {
```
It makes more sense to me to put this in `createTailnet` just before we start listening for SSH connections. The manifest will be available at that time, and the whole point is to have a stable key, so we don't need to run it more than once.

spikecurtis reviewed (commented):
Overall, looks good. Couple comments inline

spikecurtis commented on agent/agentssh/agentssh.go:
```diff
@@ -131,14 +131,7 @@ func NewServer(ctx context.Context, logger slog.Logger, prometheusRegistry *prom
-	if err != nil {
-		return nil, err
-	}
+
```
orphaned comment above

spikecurtis commented on agent/agentssh/agentssh.go:
```diff
@@ -206,7 +199,7 @@ func NewServer(ctx context.Context, logger slog.Logger, prometheusRegistry *prom
 		},
 		Handler:     s.sessionHandler,
-		HostSigners: []ssh.Signer{randomSigner},
+		HostSigners: []ssh.Signer{},
```
maybe a comment that this gets filled in just before we start listening.

mafredri reviewed (approved):
It might be worthwhile to add one more test in the `agent` package to ensure the signer is set to the expected key via the manifest, but I won't block on its addition.

Context: Right now the tests verify that the host key change works since a host key has been omitted by default, so if update isn't called, tests will fail. An unlikely scenario would be that someone added back the default key (for whatever reason), and broke the update method. No test would catch this currently, i.e. there's nothing to guard/verify the change we wanted to achieve in this PR.

ThomasK33 commented:
Adding the test was a great idea; you should have blocked the PR for its addition. 😅 

Since the `agent` package does not validate (or even skip) the SSH key validation, I've added a test to the cli ssh tests. I duplicated the `TestSSH/Stdio` test and added a host key validation, which worked fine on the first run locally, but then failed in CI and then worked again locally until it didn't. 

As it turns out, Go's stdlib is very specific about [`rsa.GenerateKey` being non-deterministic](https://github.com/golang/go/blob/5817e650946aaa0ac28956de96b3f9aa1de4b299/src/crypto/rsa/rsa.go#L272-L274), and they purposefully cut off the first byte in 50% of the cases, to ensure that trait.

Long story short, we're now computing RSA keys by hand to overcome this limitation, and we now consistently generate the same key because we're not cutting off that first byte.

On a side note, we should consider rolling out this host key validation for all CLI SSH tests in the future, but I don't want to blow the scope of this PR.
